lib/evaluator/arenaEvaluator.ts

```diff
@@ -14,21 +14,244 @@ import type { Evaluator } from "./evaluator.ts";
+ *
+ * Uses iterative processing with memoization to:
+ * 1. Avoid call stack overflow on deep structures
+ * 2. Preserve DAG structure (hash consing) to prevent memory explosion
+ * 3. Uses direct memory views instead of WASM function calls for performance
+ */
+export function fromArenaWithExports(
+  rootId: ArenaNodeId,
+  exports: Pick<
+    ArenaWasmExports,
+    "kindOf" | "symOf" | "leftOf" | "rightOf" | "debugGetArenaBaseAddr"
+  >,
+  memory?: WebAssembly.Memory,
+): SKIExpression {
+  // Get or build arena views with caching
+  let views = getOrBuildArenaViews(memory, exports);
+  const cache = new Map<number, SKIExpression>();
+  const stack: number[] = [rootId];
+
+  while (stack.length > 0) {
+    // Peek at the current node (don't pop yet, we might need to push children)
+    const id = stack[stack.length - 1];
+
+    // If we've already built this node, just pop and move on
+    if (cache.has(id)) {
+      stack.pop();
+      continue;
+    }
+
+    // Validate views are still current (arena might have grown)
+    if (views) {
+      const validatedViews = validateAndRebuildViews(views, memory, exports);
+      if (validatedViews !== views) {
+        views = validatedViews;
+      }
+    }
+
+    // Use direct memory access if views are available, otherwise fall back to WASM calls
+    const kind = views
+      ? (id < views.capacity ? views.kind[id] : exports.kindOf(id))
+      : exports.kindOf(id);
+
+    if (kind === (ArenaKind.Terminal as number)) {
+      // TERMINAL: Construct immediately and cache
+      const sym = views
+        ? (id < views.capacity ? views.sym[id] : exports.symOf(id))
+        : exports.symOf(id);
+      let expr: SKIExpression;
+      switch (sym) {
+        case ArenaSym.S:
+          expr = S;
+          break;
+        case ArenaSym.K:
+          expr = K;
+          break;
+        case ArenaSym.I:
+          expr = I;
+          break;
+        default:
+          throw new Error(`Unknown symbol tag: ${sym}`);
+      }
+      cache.set(id, expr);
+      stack.pop();
+    } else if (
+      kind === (ArenaKind.Continuation as number) ||
+      kind === (ArenaKind.Suspension as number)
+    ) {
+      // CONTINUATION/SUSPENSION: These are internal WASM-only nodes used for iterative reduction.
+      // They should never appear in the final result, but if they do (e.g., due to a bug or
+      // incomplete reduction), we cannot convert them to SKI expressions.
+      // Skip them and pop from stack to avoid infinite loops.
+      throw new Error(
+        `Cannot convert ${
+          kind === (ArenaKind.Continuation as number)
+            ? "Continuation"
+            : "Suspension"
+        } node ${id} to SKI expression. This node type is internal to the WASM reducer and should not appear in results.`,
+      );
+    } else {
+      // NON-TERMINAL: Check children
+      const leftId = views
+        ? (id < views.capacity ? views.leftId[id] : exports.leftOf(id))
```

Comment:
deref ` views.capacity` int and `views.leftId` array only once, make some `var`/`let` vars.